### ISSUE TITLE
Finalize dual trains

### ISSUE THREAD

**Issue by erik-vos** [Rails-18xx/Rails#232]
Emergency buying a dual train certificate does not work, as I discovered in 18Scan. This appears to be due to the way such trains are intialised, therefore some rework is to be done on that matter. I have split it off to a separate branch called Dual Trains.

Just a question for now:
When buying a dual (or "flippable") train card from the Pool or from another company, may that card then be flipped, i.e. buying the train on the other side than the original one? 
My rule books of 1844 and 18Scan seem to be silent on this matter. Has this ever been clarified?

**Comment by neutronc** [comment]
As far as i know the player has the right to choose which train he wants to buy on the dual train certificate. After he has decided the type the Train is finalized and the certificate cant be flipped anymore. This is at least the mechanismn on 1844. Checking 18Scan rules.

**Comment by neutronc** [comment]
18Scan is not specifiying that as well, but so the certificate is representing both train types until the player decides which type he wants and pays the respective money.

**Comment by erik-vos** [comment]
The question is: after a train card has been bought for the first time and the buyer has decided on its type, are future owners of that same card bound to that choice (i.e. when bought from the Pool or from another company). I found that 18VA rules are explicit about this: in the Pool, and only there, becomes the train type undefined again. 

As I want to prepare for that situation as well, the solution will rather deeply affect the whole train handling. One first decision to take is, whether portfolios own _trains_ or _train certificates_. I am undecided, but somewhat leaning to the latter now. On the other hand, you can always find the certificate of a certain train, so perhaps we can keep using the current usage anyhow.

**Comment by neutronc** [comment]
havent thought about that (18VA) but unless its explicitely stated the certificate becomes the train type that the player chooses upon buy from the bank for the first time. I dont want Players to be able to switch certifiate types upon crossselling them to other companies...
Thats a NO GO. buying a 1+1 and selling that as 2 Train to someone else for 100 K making 20k profit on a train sale..

**Comment by erik-vos** [comment]
OK, that's reasonable for the cross-selling case: no flipping.
And how about buying from the Pool? It may be a rare event, but we have to deal with it. From there trains must be always bought at face value. 
It would be nice to know if any other games have rules for that. I just found that in 1812 such trains do _not_ become undecided in the Pool, so that's a a different case then.

My conclusion is, that this behaviour must become configurable rather than hard-coded.

**Comment by erik-vos** [comment]
I have built in a configurable game parameter that specifies if dual trains become undecided in the Pool. Default = false.
Values:
18VA: true
18Scan: true for now, so I can test it.
1844: no for now, but I leave that to Martin who at least speaks the same language ;-)
1812 (not yet started but who knows): no
I'll ask about this matter in the 18xx mailing list.

**Comment by erik-vos** [comment]
Steve Thomas has informed me that for 18Scan and also for 1846 the setting must be true.

In the meantime I have adapted and also simplified the train creation process so that dual trains now almost work correctly. Still work to do.

However, this fix will make all older save files wherein dual trains have been bought unloadable. The train ids in these save files are incorrect. Martin, would this be a big deal for you w.r.t. 1844? If so, I could try to find a workaround, but would prefer not to do that.

**Comment by erik-vos** [comment]
That "almost work correctly" thought has turned out to get me into a quagmire of increasingly exotic errors, so I had to start afresh, and finally honor the data model (or class diagram), the basics of which I had drawn already in a early stage; see below.
![TrainsDataModel](https://user-images.githubusercontent.com/13908049/83131087-99667000-a0df-11ea-8f45-69cae44a460f.jpg)
The main news is that I have introduced the new concept of a TrainCard, which is linked to either one or two trains. It's no longer the Trains but the TrainCards that are moved between portfolios. The TrainCard knows which Train is active in case of dual trains. 

Most of the existing code outside the TrainManager realm can keep working with Trains as usual; TrainCards and Trains know each other. Most changes are caused by name changes, for instance the name TrainCertificateType has been replaced by TrainCardType.

Now that (finally) the official test suite has been passed, I have pushed the current state of affairs to my own fork as erik-vos/DualTrains, so people can have a look. Comments and questions are welcome. Before wanting to merge it with master I will first rebase my embryonic 18Scan upon DualTrains, so I can test if it really works there.
